src/librustc_codegen_llvm/back/link.rs

```diff
@@ -609,6 +594,71 @@ fn print_native_static_libs(sess: &Session, all_native_libs: &[NativeLibrary]) {
     }
 }
 
+pub fn linker_and_flavor(sess: &Session) -> Result<(PathBuf, LinkerFlavor), ()> {
+    fn from<F>(
+        sess: &Session,
+        linker: Option<PathBuf>,
+        flavor: Option<LinkerFlavor>,
+        otherwise: F,
+    ) -> Result<(PathBuf, LinkerFlavor), ()>
+    where
+        F: FnOnce() -> Result<(PathBuf, LinkerFlavor), ()>
+    {
+        match (linker, flavor) {
+            (Some(linker), Some(flavor)) => Ok((linker, flavor)),
+            // only the linker flavor is known; use the default linker for the selected flavor
+            (None, Some(flavor)) => Ok((PathBuf::from(match flavor {
+                LinkerFlavor::Em => "emcc",
+                LinkerFlavor::Gcc => "gcc",
+                LinkerFlavor::Ld => "ld",
+                LinkerFlavor::Msvc => "link.exe",
+                LinkerFlavor::Lld(_) => "lld",
+            }), flavor)),
+            // infer the linker flavor from the linker name
+            (Some(linker), None) => {
+                let stem = linker.file_stem().and_then(|stem| stem.to_str()).ok_or_else(|| {
+                    sess
+                        .struct_err(&format!("couldn't extract file stem from specified linker"))
+                        .emit();
+                })?.to_owned();
+
+                let flavor = if stem == "emcc" {
+                    LinkerFlavor::Em
+                } else if stem == "gcc" || stem.ends_with("-gcc") {
+                    LinkerFlavor::Gcc
+                } else if stem == "ld" || stem == "ld.lld" || stem.ends_with("-ld") {
+                    LinkerFlavor::Ld
+                } else if stem == "link" || stem == "lld-link" {
+                    LinkerFlavor::Msvc
+                } else {
+                    sess
+                        .struct_err(&format!("couldn't infer linker flavor from specified linker"))
+                        .emit();
+                    return Err(());
+                };
+
+                Ok((linker, flavor))
+            },
+            (None, None) => otherwise(),
```

Comment:
Stylistically I might recommend instead of a cps style to instead do something like:

* Emit only fatal errors here via `sess` (it doesn't look like the error is ever "caught"?)
* Return `Option` instead of `Result`
* Instead do something like:

```rust
if let Some(ret) = locate_linker(&sess.opts) {
    return Some(ret)
}
if let Some(ret) = locate_linker(&sess.target.target) {
    return Some(ret)
}
sess.fatal("...")
```